Assets/Scripts/PlayerSupervisor.cs

```diff
@@ -1,317 +1,385 @@
+
+    public int boundaryReboundLimit = 10;
+
+    [Tooltip("How often to check for anomalies (0 eliminates check)")]
+    [Range(0f, 2f)]
+    public float detectionFreq = 1f;
+
+    [Tooltip("Angle that ball will ricochet off ceiling to prevent juggling")]
+    [Range(0f, 5f)]
+    public float ceilingReboundAngle = 0f;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        gameManager = FindObjectOfType<GameManager>();
+        
+        if (!playerAgent)
+            playerAgent = FindObjectOfType<PlayerAgent>();
+
+        // the code will check whether or not to execute
+        // based on the block.name assigned in the Inspector Window
+        // in the RandomBlockCreator empty child object
+        randomBlockCreator = FindObjectOfType<RandomBlockCreator>();
+        if (randomBlockCreator)
+            randomBlockCreator.setupBlocks();
+        
+        if (!ball)
+            ball = FindObjectOfType<Ball>();
+        ballOffset = ball.transform.localPosition;
+
+        // Performance tracking - has to come before countblocks
+        if(!isMultiTraining && gameManager.trackingPerformanceTF)
+            dataManager = FindObjectOfType<PerformanceDataManager>();
+        
+        if (!paddle)
+            paddle = FindObjectOfType<Paddle>();
+        paddleOffset = paddle.transform.localPosition;
+
+        // Check if scene is ready for training
+        if (gameManager.trainingMode)
+            ResetState();
+        else
+            CountBlocks();
+    }
+
+    void CountBlocks()
+    {
+        activeBlocks = 0;
+
+        // TODO need to count only blocks in each player's environment to
+        //  work for general multiplayer use; Maybe keep all blocks in a
+        //  container for easier tracking
+        if (gameManager.trainingMode)
+        {
+            foreach(Transform child in trainingBlocksInstance.transform)
+            {
+                if (child.gameObject.GetComponent<Block>() && child.gameObject.activeSelf)
+                    ++activeBlocks;
+            }
+        }
+        else
+        {
+            foreach (Block block in FindObjectsOfType<Block>())
+            {
+                if (block.gameObject.activeSelf)
+                    ++activeBlocks;
+            }
+        }
+        if(!isMultiTraining && gameManager.trackingPerformanceTF)
+            dataManager.SetStartingNumBlocks(activeBlocks);
+    }
+
+    public void PlayerReady()
+    {
+        if (playerState == PlayerState.Waiting)
+        {
+            playerState = PlayerState.Ready;
+            gameManager.StartGame(this);
+        }
+    }
+
+    public void StartGame()
+    {
+        if (gameManager.trainingMode && !trainingBlocks)
+        {
+            Debug.LogError("trainingBlocks reference missing");
+            return;
+        }
+
+        if(!isMultiTraining && gameManager.trackingPerformanceTF)
+            dataManager.SetStartingNumBlocks(activeBlocks);
```

Comment:
This is the line I mentioned in the above comment. I think this is the most appropriate place to be calling SetStartingNumBlocks so the one above can be removed.